docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,341 @@
+
+The database stores the data retrieved from the full node and provide this data for the RPC server instance. Since the frequency
+that blocks are generated on the chain are in the range from 5 seconds to 7 seconds on average, the _write_ back pressure is not
+very high from a modern database perspective. While the frequency and number of requests for reading the data from the database will
+be much larger due to the fact that the RPC service instance can be scaled. Therefore, a database that provides a high read
+throughput should be favored.
+
+For this is initial solution implementation it is proposed that the relational database [Postgresql](https://www.postgresql.org/) should be used in order to support
+the RPC scalability and this will also provide more flexibility when implementing the RPC v2 endpoints that can return data
+in different forms and database indexes might also be leveraged in order to boost the query responses performance.
+
+The data needs to be available both for the Ingest Service (writes) and the RPC server instance (reads) and these service might be running from different machines so an embedded database
+is not recommended in this case since accessing the data remotely might not be optimal for an embedded key-value database. Also
+since the RPC might have many server instances (or processes) running that will need to retrieve data concurrently it is recommended to use
+a well-known robust database engine that can support such a load.
+
+Also, a database that can support ACID transactions is important to provide more guarantees that
+the data was successfully inserted in the database and that an acknowledgement can be sent back to the Ingest Service to notify the
+full node to prune the inserted data.
+
+#### [Database Schema](#database-schema)
+
+One of the challenges when implementing this solution is how to design a database schema that can be suitable to return responses that are
+equivalent to the existing endpoints but at the same time offers flexibility in returning customized responses in the future.
+
+Currently, CometBFT uses an abstraction layer from [cometbft-db](https://github.com/cometbft/cometbft-db) in order to support multiple embedded databases. These databases store
+data as key-value pairs using a byte array datatype, for example to set a value for a key:
+
+```go
+func (db *GoLevelDB) Set(key []byte, value []byte) error {
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+	if value == nil {
+		return errValueNil
+	}
+	if err := db.db.Put(key, value, nil); err != nil {
+		return err
+	}
+	return nil
+}
+```
+
+Since the RPC Companion stores the information in a relational database, there are opportunities to better structure and normalize the data. Here is the schema definition for a table to persist a normalized `ResultBlock` data structure in the database (Postgresql)
+
+```sql
+-- Table: comet.result_block
+
+CREATE TABLE IF NOT EXISTS comet.result_block
+(
+    block_id_hash bytea NOT NULL,
```

Comment:
Shouldn't we be using foreign keys for this kind of fields?

Reply:
the schema is still evolving, this is the latest schema and it might change. The comments on schema are helpful but I'd think they would be worth it to have in the PR for the rpc-companion when it's submitted.